File: apps/server/src/routes/games.ts
import { zValidator } from "@hono/zod-validator";
import { and, desc, eq, inArray, lt, sql } from "drizzle-orm";
import { Hono } from "hono";
import { gamesQuerySchema } from "@aoe4-almanac/shared";
import type { AppContext } from "../auth/middleware.ts";
import { db, sqlite } from "../db/client.ts";
import { gameParticipants, games } from "../db/schema.ts";

export const gamesRoutes = new Hono<AppContext>();

gamesRoutes.get("/", zValidator("query", gamesQuerySchema), (c) => {
  const userId = c.get("userId");
  const q = c.req.valid("query");

  const where = [eq(games.userId, userId)];
  if (q.civ) where.push(eq(games.myCivSlug, q.civ));
  if (q.map) where.push(eq(games.mapSlug, q.map));
  if (q.result) where.push(eq(games.myResult, q.result));
  if (q.kind) where.push(eq(games.kind, q.kind));
  if (q.leaderboard) where.push(eq(games.leaderboard, q.leaderboard));
  if (q.since) where.push(sql`${games.startedAt} >= ${q.since}`);
  if (q.cursor) where.push(lt(games.id, q.cursor));

  // opp_civ filter requires a subquery over participants
  let candidateIds: number[] | null = null;
  if (q.opp_civ) {
    candidateIds = (
      sqlite()
        .prepare(
          `SELECT DISTINCT g.id AS id
           FROM games g
           JOIN game_participants p ON p.game_id = g.id
           WHERE g.user_id = ? AND p.is_self = 0 AND p.civ_slug = ?`,
        )
        .all(userId, q.opp_civ) as { id: number }[]
    ).map((r) => r.id);
    if (candidateIds.length === 0) {
      return c.json({ games: [], next_cursor: null });
    }
    where.push(inArray(games.id, candidateIds));
  }

  const rows = db()
    .select()
    .from(games)
    .where(and(...where))
    .orderBy(desc(games.startedAt), desc(games.id))
    .limit(q.limit)
    .all();

  const ids = rows.map((r) => r.id);
  const parts =
    ids.length > 0
      ? db()
          .select()
          .from(gameParticipants)
          .where(inArray(gameParticipants.gameId, ids))
          .all()
      : [];
  const partsByGame = new Map<number, typeof parts>();
  for (const p of parts) {
    const arr = partsByGame.get(p.gameId) ?? [];
    arr.push(p);
    partsByGame.set(p.gameId, arr);
  }

  const out = rows.map((g) => ({
    ...g,
    is_variant: undefined,
    participants: (partsByGame.get(g.id) ?? []).map((p) => ({
      id: p.id,
      game_id: p.gameId,
      team: p.team,
      is_self: Boolean(p.isSelf),
      profile_id: p.profileId,
      name: p.name,
      civ_slug: p.civSlug,
      civ_randomized: p.civRandomized,
      result: p.result,
      rating: p.rating,
      rating_diff: p.ratingDiff,
      mmr: p.mmr,
    })),
    aoe4world_game_id: g.aoe4worldGameId,
    started_at: g.startedAt,
    duration_seconds: g.durationSeconds,
    map_slug: g.mapSlug,
    user_id: g.userId,
    my_team: g.myTeam,
    my_civ_slug: g.myCivSlug,
    my_civ_randomized: g.myCivRandomized,
    my_result: g.myResult,
    my_rating: g.myRating,
    my_rating_diff: g.myRatingDiff,
    my_mmr: g.myMmr,
    imported_at: g.importedAt,
    created_at: g.createdAt,
    updated_at: g.updatedAt,
  }));

  const nextCursor =
    rows.length === q.limit && rows[rows.length - 1]
      ? (rows[rows.length - 1]?.id ?? null)
      : null;

  return c.json({ games: out, next_cursor: nextCursor });
});

gamesRoutes.get("/:id", (c) => {
  const userId = c.get("userId");
  const id = Number(c.req.param("id"));
  if (!Number.isFinite(id)) return c.json({ error: "bad id" }, 400);
  const row = db()
    .select()
    .from(games)
    .where(and(eq(games.id, id), eq(games.userId, userId)))
    .get();
  if (!row) return c.json({ error: "not found" }, 404);
  const parts = db()
    .select()
    .from(gameParticipants)
    .where(eq(gameParticipants.gameId, id))
    .all();
  return c.json({
    ...row,
    aoe4world_game_id: row.aoe4worldGameId,
    started_at: row.startedAt,
    duration_seconds: row.durationSeconds,
    map_slug: row.mapSlug,
    user_id: row.userId,
    my_team: row.myTeam,
    my_civ_slug: row.myCivSlug,
    my_civ_randomized: row.myCivRandomized,
    my_result: row.myResult,
    my_rating: row.myRating,
    my_rating_diff: row.myRatingDiff,
    my_mmr: row.myMmr,
    imported_at: row.importedAt,
    raw_payload_json: row.rawPayloadJson,
    raw_payload: row.rawPayloadJson ? JSON.parse(row.rawPayloadJson) : null,
    created_at: row.createdAt,
    updated_at: row.updatedAt,
    participants: parts.map((p) => ({
      id: p.id,
      game_id: p.gameId,
      team: p.team,
      is_self: Boolean(p.isSelf),
      profile_id: p.profileId,
      name: p.name,
      civ_slug: p.civSlug,
      civ_randomized: p.civRandomized,
      result: p.result,
      rating: p.rating,
      rating_diff: p.ratingDiff,
      mmr: p.mmr,
    })),
  });
});

File: src-tauri/src/commands/games.rs
use crate::db::{self, Db};
use crate::error::{AppError, AppResult};
use rusqlite::{params, params_from_iter, ToSql};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use tauri::State;

#[derive(Deserialize, Default)]
pub struct GamesQuery {
    pub civ: Option<String>,
    pub map: Option<String>,
    pub result: Option<String>,
    pub kind: Option<String>,
    pub leaderboard: Option<String>,
    pub since: Option<i64>,
    pub cursor: Option<i64>,
    pub opp_civ: Option<String>,
    pub limit: Option<i64>,
}

#[derive(Serialize, Clone)]
pub struct Participant {
    pub id: i64,
    pub game_id: i64,
    pub team: i64,
    pub is_self: bool,
    pub profile_id: Option<i64>,
    pub name: String,
    pub civ_slug: String,
    pub civ_randomized: Option<i64>,
    pub result: String,
    pub rating: Option<i64>,
    pub rating_diff: Option<i64>,
    pub mmr: Option<i64>,
}

fn fetch_participants(
    conn: &rusqlite::Connection,
    game_ids: &[i64],
) -> AppResult<HashMap<i64, Vec<Participant>>> {
    if game_ids.is_empty() {
        return Ok(HashMap::new());
    }
    let placeholders = game_ids.iter().map(|_| "?").collect::<Vec<_>>().join(",");
    let sql = format!(
        "SELECT id, game_id, team, is_self, profile_id, name, civ_slug,
                civ_randomized, result, rating, rating_diff, mmr
         FROM game_participants WHERE game_id IN ({})",
        placeholders
    );
    let mut stmt = conn.prepare(&sql)?;
    let params: Vec<&dyn ToSql> = game_ids.iter().map(|id| id as &dyn ToSql).collect();
    let parts: Vec<Participant> = stmt
        .query_map(params_from_iter(params), |r| {
            Ok(Participant {
                id: r.get(0)?,
                game_id: r.get(1)?,
                team: r.get(2)?,
                is_self: r.get::<_, i64>(3)? != 0,
                profile_id: r.get(4)?,
                name: r.get(5)?,
                civ_slug: r.get(6)?,
                civ_randomized: r.get(7)?,
                result: r.get(8)?,
                rating: r.get(9)?,
                rating_diff: r.get(10)?,
                mmr: r.get(11)?,
            })
        })?
        .collect::<Result<_, _>>()?;
    let mut by_game: HashMap<i64, Vec<Participant>> = HashMap::new();
    for p in parts {
        by_game.entry(p.game_id).or_default().push(p);
    }
    Ok(by_game)
}

fn game_to_json(row: &rusqlite::Row, col_names: &[String]) -> rusqlite::Result<serde_json::Value> {
    let mut obj = serde_json::Map::new();
    for (i, name) in col_names.iter().enumerate() {
        let v: rusqlite::types::Value = row.get(i)?;
        obj.insert(name.clone(), value_to_json(v));
    }
    Ok(serde_json::Value::Object(obj))
}

fn value_to_json(v: rusqlite::types::Value) -> serde_json::Value {
    use rusqlite::types::Value;
    match v {
        Value::Null => serde_json::Value::Null,
        Value::Integer(i) => serde_json::Value::from(i),
        Value::Real(f) => serde_json::Value::from(f),
        Value::Text(s) => serde_json::Value::String(s),
        Value::Blob(b) => serde_json::Value::String(format!("<blob:{}>", b.len())),
    }
}

#[tauri::command]
pub fn games_list(
    state: State<'_, Db>,
    query: Option<GamesQuery>,
) -> AppResult<serde_json::Value> {
    let conn = state.0.lock().expect("db poisoned");
    let user_id = db::local_user_id(&conn)?;
    let q = query.unwrap_or_default();
    let limit = q.limit.unwrap_or(50).clamp(1, 500);

    let mut where_parts = vec!["user_id = ?".to_string()];
    let mut params_vec: Vec<Box<dyn ToSql>> = vec![Box::new(user_id)];

    if let Some(v) = &q.civ {
        where_parts.push("my_civ_slug = ?".into());
        params_vec.push(Box::new(v.clone()));
    }
    if let Some(v) = &q.map {
        where_parts.push("map_slug = ?".into());
        params_vec.push(Box::new(v.clone()));
    }
    if let Some(v) = &q.result {
        where_parts.push("my_result = ?".into());
        params_vec.push(Box::new(v.clone()));
    }
    if let Some(v) = &q.kind {
        where_parts.push("kind = ?".into());
        params_vec.push(Box::new(v.clone()));
    }
    if let Some(v) = &q.leaderboard {
        where_parts.push("leaderboard = ?".into());
        params_vec.push(Box::new(v.clone()));
    }
    if let Some(v) = q.since {
        where_parts.push("started_at >= ?".into());
        params_vec.push(Box::new(v));
    }
    if let Some(v) = q.cursor {
        where_parts.push("id < ?".into());
        params_vec.push(Box::new(v));
    }

    if let Some(opp) = &q.opp_civ {
        let mut stmt = conn.prepare(
            "SELECT DISTINCT g.id FROM games g
             JOIN game_participants p ON p.game_id = g.id
             WHERE g.user_id = ?1 AND p.is_self = 0 AND p.civ_slug = ?2",
        )?;
        let ids: Vec<i64> = stmt
            .query_map(params![user_id, opp], |r| r.get::<_, i64>(0))?
            .collect::<Result<_, _>>()?;
        drop(stmt);
        if ids.is_empty() {
            return Ok(serde_json::json!({ "games": [], "next_cursor": null }));
        }
        let placeholders = ids.iter().map(|_| "?").collect::<Vec<_>>().join(",");
        where_parts.push(format!("id IN ({})", placeholders));
        for id in &ids {
            params_vec.push(Box::new(*id));
        }
    }

    params_vec.push(Box::new(limit));
    let sql = format!(
        "SELECT * FROM games WHERE {}
         ORDER BY started_at DESC, id DESC LIMIT ?",
        where_parts.join(" AND ")
    );

    let mut stmt = conn.prepare(&sql)?;
    let col_names: Vec<String> = stmt.column_names().iter().map(|s| s.to_string()).collect();
    let rows: Vec<serde_json::Value> = stmt
        .query_map(params_from_iter(params_vec.iter().map(|b| b.as_ref())), |r| {
            game_to_json(r, &col_names)
        })?
        .collect::<Result<_, _>>()?;

    let ids: Vec<i64> = rows.iter().filter_map(|r| r.get("id").and_then(|v| v.as_i64())).collect();
    let by_game = fetch_participants(&conn, &ids)?;

    let games: Vec<serde_json::Value> = rows
        .into_iter()
        .map(|mut r| {
            if let Some(id) = r.get("id").and_then(|v| v.as_i64()) {
                let parts = by_game.get(&id).cloned().unwrap_or_default();
                if let Some(obj) = r.as_object_mut() {
                    obj.insert("participants".into(), serde_json::to_value(parts).unwrap());
                }
            }
            r
        })
        .collect();

    let next_cursor = if games.len() as i64 == limit {
        games.last().and_then(|g| g.get("id").and_then(|v| v.as_i64()))
    } else {
        None
    };

    Ok(serde_json::json!({ "games": games, "next_cursor": next_cursor }))
}

#[tauri::command]
pub fn games_get(state: State<'_, Db>, id: i64) -> AppResult<serde_json::Value> {
    let conn = state.0.lock().expect("db poisoned");
    let user_id = db::local_user_id(&conn)?;

    let mut stmt = conn.prepare("SELECT * FROM games WHERE id = ?1 AND user_id = ?2")?;
    let col_names: Vec<String> = stmt.column_names().iter().map(|s| s.to_string()).collect();
    let mut row = stmt
        .query_row(params![id, user_id], |r| game_to_json(r, &col_names))
        .map_err(|e| match e {
            rusqlite::Error::QueryReturnedNoRows => AppError::msg("not found"),
            other => AppError::from(other),
        })?;

    let by_game = fetch_participants(&conn, &[id])?;
    if let Some(obj) = row.as_object_mut() {
        // Expose decoded raw_payload alongside the raw JSON string.
        let parsed: Option<serde_json::Value> = obj
            .get("raw_payload_json")
            .and_then(|v| v.as_str())
            .and_then(|s| serde_json::from_str(s).ok());
        obj.insert("raw_payload".into(), parsed.unwrap_or(serde_json::Value::Null));
        obj.insert(
            "participants".into(),
            serde_json::to_value(by_game.get(&id).cloned().unwrap_or_default()).unwrap(),
        );
    }
    Ok(row)
}
